core/src/main/scala/org/broadinstitute/dsde/rawls/entities/base/InputExpressionReassembler.scala

```diff
@@ -169,17 +175,61 @@ object InputExpressionReassembler {
+    def updateInputExprValueMapForObjectInput(inputExprWithEvaluatedRef: JsValue, objectFieldsFromInput: mutable.Seq[ValueTypeObjectFieldTypes]): JsObject = {
+      val updatedObject: Map[String, JsValue] = inputExprWithEvaluatedRef.asJsObject.fields.map { case (inputName, evaluatedValue) =>
+        val inputObjectField: Option[ValueTypeObjectFieldTypes] = objectFieldsFromInput.find(x => x.getFieldName == inputName)
+
+        val updatedValue: JsValue = inputObjectField match {
+          case Some(objectField) => objectField.getFieldType.getTypeName match {
+            case TypeNameEnum.OBJECT =>
+              // if there are nested objects, recursively call this method to reach leaf fields
+              updateInputExprValueMapForObjectInput(evaluatedValue, objectField.getFieldType.getObjectFieldTypes.asScala)
+            case TypeNameEnum.ARRAY => evaluatedValue match {
+              case JsNull => JsArray()
+              case JsArray(_) => evaluatedValue
+              case _ => JsArray(evaluatedValue)
+            }
+            case _ => evaluatedValue
+          }
+          case None => evaluatedValue
+        }
+        inputName -> updatedValue
+      }
+
+      JsObject(updatedObject)
+    }
+
     // when there are no root entities handle as a single root entity with empty string for name
     rootEntityNames.getOrElse(Seq("")).map { entityName =>
       // in the case of literal JSON there are no LookupExpressions
       val evaluatedLookupMapTry = mapOfEntityToEvaluatedExprMap.getOrElse(entityName, Success(Map.empty[LookupExpression, JsValue]))
-      val inputExprWithEvaluatedRef = evaluatedLookupMapTry.map { lookupMap =>
+      val inputExprWithEvaluatedRef: Try[JsValue] = evaluatedLookupMapTry.map { lookupMap =>
         val visitor = new ReconstructExpressionVisitor(lookupMap)
         visitor.visit(parsedTree)
       }
 
-      entityName -> inputExprWithEvaluatedRef
+      /* In SlickExpressionEvaluator, for attribute references that evaluate to AttributeValueList type, the elements inside the list are extracted and returned.
+         As a result, for attributes that contain only 1 element, we can no longer identify whether the original value was a list or a single AttributeValue.
+         Because of this type erasure, when we are reconstructing object type inputs, List/Array with single element get incorrectly reconstructed as single AttributeValue,
+         which leads to Cromwell failing while processing inputs (https://broadworkbench.atlassian.net/browse/BW-678).
+         To fix this, once the object type inputs have been reconstructed, we iterate over each field for object input and verify for Array type inputs that the
+         evaluated value is also an Array. If it isn't we convert it into Array and updated the evaluated object.
+         Note: This approach does not fix the problem if there is single element array inside an array. This would be fixed in https://broadworkbench.atlassian.net/browse/BW-846.
```

Comment:
Or you could have an equivalent, recursive `updateInputExprValueMapForArrayInput`, presumably?

Reply:
Ah I see what you mean. Yeah looking inside Array inputs might be tricky especially for deeper nested arrays. I would hope if we look at this bug again, we would solve the problem for real by fixing the type erasure.